README.md
# KPP-Mistra

KPP-Mistra is a version of the KPP software customized for the
[Mistra model](https://github.com/MistraModel/Mistra).

KPP-Mistra should only be used in conjunction with the
[Mistra model](https://github.com/MistraModel/Mistra); for any other
application use the [original version](https://people.cs.vt.edu/~asandu/Software/Kpp/)
or check out the GEOS-Chem [community version](https://github.com/KineticPreProcessor/KPP).

******************************************************************************

## KPP

KPP is a symbolic chemistry Kinetics PreProcessor, developed by
V. Damian and A. Sandu, with the contribution of R. Sander. KPP builds
simulation code for chemical kinetic systems. For more information, go
to the [KPP website](https://people.cs.vt.edu/~asandu/Software/Kpp/).

Copyright (C) 1995-1997 Valeriu Damian and Adrian Sandu, CGRER, Univ. Iowa (USA).

Copyright (C) 1997-2016 Adrian Sandu, Michigan Tech. & Virginia Tech. (USA),
with contributions from Rolf Sander, Max-Planck Institute for Chemistry, Mainz (Germany).

KPP is free software; you can redistribute it and/or modify it under the
terms of the GNU General Public License as published by the Free
Software Foundation (http://www.gnu.org/copyleft/gpl.html); either
version 2 of the License, or (at your option) any later version.

KPP is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
for more details.

A copy of the GNU General Public License can be found in the `LICENCE`
file; you can also consult http://www.gnu.org/copyleft/gpl.html or
write to the Free Software Foundation, Inc., 59 Temple Place - Suite
330, Boston, MA 02111-1307, USA.

Adrian Sandu  
Computer Science Department  
Virginia Polytechnic Institute and State University  
Blacksburg, VA 24060, USA  
E-mail: sandu@cs.vt.edu

### Reference

V. Damian, A. Sandu, M. Damian, F. Potra, G.R. Carmichael: "The
Kinetic PreProcessor KPP -- A Software Environment for Solving
Chemical Kinetics", Computers and Chemical Engineering, 26, 11,
1567-1579, 2002 [[link](https://doi.org/10.1016/S0098-1354(02)00128-X)].

******************************************************************************

## KPP-Mistra

KPP-Mistra was developed from *KPP v2.2.3* and includes all the bugfixes
and modifications made by R. Sander and J. Bock (see the `f77` and
`mistra` branches at: https://bitbucket.org/gcst/kpp), plus additional
modifications specific to Mistra made by J. Bock. A summary of the
changes can be found in the `CHANGELOG.md` file.

The stable version of KPP-Mistra can be downloaded from the
[Releases](https://github.com/MistraModel/KPP-Mistra/releases)
page.

### Instructions

#### To install KPP-Mistra:

1. Check that FLEX (open-source lexical analizer) is installed using
   the command: `flex --version`

2. Note the path of the FLEX library (`libfl.a` or `libfl.so`).

3. Check that BISON (open-source parser generator) is installed using
   the command: `bison --version`

4. Define the `$KPP_HOME` environment variable to point to the location of KPP-Mistra.
   If, for example, KPP-Mistra is installed in `$HOME/KPP-Mistra`:

  - if using the **C** shell (or **tcsh** shell), add the following to
    `$HOME/.cshrc` (or `$HOME/.tcshrc`):
    ```shell
    setenv KPP_HOME $HOME/KPP-Mistra
    set path=( $path $KPP_HOME/bin )
    ```

  - if using the **bash** shell, add the following to `$HOME/.bashrc`:
    ```shell
    export KPP_HOME=$HOME/KPP-Mistra
    export PATH=${PATH}:$KPP_HOME/bin
    ```

   - Execute the command `source ~/.cshrc` (or `.tcshrc`, or `.bashrc`) to make
     sure that these changes are in effect. Alternatively, close and reopen the terminal.

5. In `$KPP_HOME` edit the file `Makefile.defs` and follow the instructions to
   specify the compiler, the location of the FLEX library, etc...

6. In the `KPP_HOME` directory build the source files using the command: `make`

#### To clean KPP-Mistra:

1. Delete the KPP object files with the command: `make clean`

2. Delete the KPP object files and binaries with the command: `make distclean`

#### To use KPP-Mistra:

The *KPP User's Manual* can be found in the `doc/` directory.

Please refer to the *Mistra manual* in the `doc/` directory of the
[Mistra repository](https://github.com/MistraModel/Mistra) for
further information.


CHANGELOG.md
Changes to KPP-2.2.3, for use with the Mistra model
===================================================

Author: J. Bock (josue.bock@meteo.fr)

1. In main directory:
   1. create `bin/` directory.
   2. edit `Makefile.defs` according to KPP instructions.

2. In `src/` directory:
   1. edit `code_f77.c` and decrease the number of written columns (integer case) from 12 to 11 (or less):
      - `maxCols = 11`
      - `if (((i+1) % 11 == 0 ...`
      - comments starting with `!` instead of `C` (optional, unactivated)
   2. `code_f90.c` from R. Sander's version.
   3. edit `gdata.h`, increase `MAX_EQN` (5000, or more) and `MAX_SPECIES` (3000, or more).
      After R. Sander changes: also increase `MAX_K` (300).
   4. use `scan.l` from R. Sander's version (dimensions increased to 300, related to change in `gdata.h` here).
   5. edit `scan.y` and increase the dimension of the character variable `str` from 80 to 300.
   6. correct `gen.c` subroutine `GenerateJacReactantProd : allocate integer arrays instead of a declaration`.

3. In `int/` directory, use corrected version of `int/rosenbrock.f`:
   1. `ode_Fun` is not the correct name, replaced by `FunTemplate`.
   2. `ode_Jac` is not the correct name, replaced by `JacTemplate`.

4. In `util/` directory:
   1. use empty version of `UserRateLaws.f`.

Further changes to KPP-2.2.3, after `Forcheck`-ing the code
-----------------------------------------------------------

In `int/rosenbrock.f` file:
1. remove all calls to subroutine `WCOPY`.
2. remove all calls to subroutine `WSCAL`.


Update: 5 March 2017 (J.B.)
===========================

- Bugfix in `src/code_f77.c`: in `F77_DeclareData`, a bug arise if the number of species (or reactions) is a multiple of `max_lines` corrected by adding `if( min < max )`.
- In the same source file: dynamic column filling, to take 3-5 digits into account, and avoid too long files (and too numerous continuation lines).
- Cleaning of the whole source file, according to compilation messages.
- In `src/gen.c`: bugfix to print correct file extension when running KPP, plus cleaning of this file according to compilation messages.
- In `util/sutil.f`: bugfix in subroutine `KppDecomp`: there was already a test to prevent `JVS(...) == 0`, but there were 2 mistakes:
  1. only `nvar` values were tested, not covering the whole `JVS(:)` values.
  2. a strict equality (`== 0`) was used, while `abs < tiny` is safer.


Update: April 2019
==================

All changes listed above have been applied to KPP v2.2.3 and documented with Pull Requests #1-#5 and #9
in the __KPP-Mistra__ repository (https://github.com/MistraModel/KPP-Mistra).

Additional changes by P. Bräuer & R. Sommariva:
- Move license file to main directory and delete `gpl/` directory.
- Update and format `README.md` and `CHANGELOG.md` files.
- Delete file `diff_version_rolf_sander.txt`.
- Delete directory `int.modified_WCOPY/` (backup/old version of `int/`?).


Update: June 2021 (R.S.)
========================

- Correct file permissions (remove execute).
- Create `.gitignore` files.
- Update `README.md` and `CHANGELOG.md` files.


Update: July 2022 (R.S.)
========================

- Revert some changes to the original KPP.
- Remove commented code and garbage files.
- Include additional Mistra-specific changes by J. Bock, resolve compilation warnings.
- Add/update `.gitignore` files.
- Rename license file to `LICENSE`.
- Update `README.md` and `CHANGELOG.md` files.


src/kpp.c
/******************************************************************************

  KPP - The Kinetic PreProcessor
        Builds simulation code for chemical kinetic systems

  Copyright (C) 1995-1996 Valeriu Damian and Adrian Sandu
  Copyright (C) 1997-2005 Adrian Sandu

  KPP is free software; you can redistribute it and/or modify it under the
  terms of the GNU General Public License as published by the Free Software
  Foundation (http://www.gnu.org/copyleft/gpl.html); either version 2 of the
  License, or (at your option) any later version.

  KPP is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
  details.

  You should have received a copy of the GNU General Public License along
  with this program; if not, consult http://www.gnu.org/copyleft/gpl.html or
  write to the Free Software Foundation, Inc., 59 Temple Place - Suite 330,
  Boston, MA  02111-1307,  USA.

  Adrian Sandu
  Computer Science Department
  Virginia Polytechnic Institute and State University
  Blacksburg, VA 24060
  E-mail: sandu@cs.vt.edu

******************************************************************************/


#include <stdlib.h>
#include <string.h>
#include "gdata.h"
#include "scan.h"

char *eqFileName;
char *rootFileName = "ff";
char Home[ MAX_PATH ] = "";

short int linStru[ MAX_SPECIES ];
short int colStru[ MAX_SPECIES ];
short int bestStru[ MAX_SPECIES ];
short int *Stru;

enum stru_criteria { UNSORT, LINSORT, COLSORT, BESTSORT };

void EqCopy( EQ_VECT e1, EQ_VECT e2 )
{
int i;

  for( i = 0; i < EqnNr; i++ ) e2[i] = e1[i];
}

int NoSort( const void *p1, const void *p2 )
{
  return -1;
}

int CodeCmp( const void *p1, const void *p2 )
{
CODE *c1, *c2;

  c1 = (CODE*)p1;
  c2 = (CODE*)p2;

  if ( *c1 < *c2 ) return -1;
  if ( *c1 > *c2 ) return 1;
  return 0;
}

int CodeRCmp( const void *p1, const void *p2 )
{
int rc1, rc2;
CODE *c1, *c2;

  c1 = (CODE*)p1;
  c2 = (CODE*)p2;

  rc1 = Reactive[ ReverseCode[ *c1 ] ];
  rc2 = Reactive[ ReverseCode[ *c2 ] ];
  if ( rc1 > rc2 ) return -1;
  if ( rc1 < rc2 ) return 1;
  if ( *c1 < *c2 ) return -1;
  if ( *c1 > *c2 ) return 1;
  return 0;
}

int CodeSSCmp( const void *p1, const void *p2 )
{
  return -CodeRCmp(p1,p2);
}

int CodeSCmp( const void *p1, const void *p2 )
{
CODE *c1, *c2;
short int sc1, sc2;

  c1 = (CODE*)p1;
  c2 = (CODE*)p2;

  sc1 = Stru[ ReverseCode[ *c1 ] ];
  sc2 = Stru[ ReverseCode[ *c2 ] ];

  if ( sc1 > sc2 ) return 1;
  if ( sc1 < sc2 ) return -1;
  if ( *c1 < *c2 ) return 1;
  if ( *c1 > *c2 ) return -1;
  return 0;
}

void UpdateStructJ()
{
int i,j,k;

  for ( i=0; i<VarNr; i++ )
    for ( j=0; j<VarNr; j++ )
      structJ[i][j]=(i==j)?1:0;

  for (i = 0; i < VarNr; i++)
    for (j = 0; j < VarNr; j++)
      for (k = 0; k < EqnNr; k++)
        if ( Stoich[i][k]*((Stoich_Left[j][k])?1:0) != 0.0 )
          structJ[i][j]=1;

  for ( i=0; i<VarNr; i++ )
    for ( j=0; j<VarNr; j++ )
      LUstructJ[i][j]=structJ[i][j];

}

int ComputeLUStructJ()
{
int i,j,k;
int nu,nl;

  for (j = 0; j < VarNr-1; j++) {
    for (i = j+1; i < VarNr; i++) {
      if( LUstructJ[i][j] ) {
        for (k = j; k < VarNr; k++)
         /*   LUstructJ[i][k] += LUstructJ[j][k]; */
	   if ( LUstructJ[j][k] != 0 )
              LUstructJ[i][k] = 1;
      }
    }
  }


  nu = 0; nl = 0;
  for (i = 0; i < VarNr; i++)
    for (j = 0; j < VarNr; j++)
      if( LUstructJ[i][j] ) {
        if(i > j) nl++;
        if(i <= j) nu++;
      }

  return nu+nl;
}

int LUnonZero()
{
CODE v[MAX_SPECIES];
CODE *var;
int i,j;

  var = v;
  if( Stru != bestStru ) {
    for( i=0; i<VarNr; i++ )
      var[i] = Code[i];
    qsort( (void*)var, VarNr, sizeof(CODE), CodeSCmp );
  } else {
    var = bestStru;
  }

  for (i = 0; i < VarNr; i++)
    for (j = 0; j < VarNr; j++)
      LUstructJ[i][j] = structJ[ ReverseCode[var[i]] ][ ReverseCode[var[j]] ];

  return ComputeLUStructJ();
}

void LinColSparsity()
{
int i,j,k;
int nlin, ncol;

  for ( i=0; i<VarNr; i++ )
    for ( j=0; j<VarNr; j++ )
      structJ[i][j]=(i==j)?1:0;

  for (i = 0; i < VarNr; i++)
    for (j = 0; j < VarNr; j++)
      for (k = 0; k < EqnNr; k++)
        if ( Stoich[i][k]*((Stoich_Left[j][k])?1:0) != 0.0 )
          structJ[i][j]=1;

  for ( i=0; i<VarNr; i++ ) {
    linStru[i] = 0;
    for (j = 0; j < VarNr; j++)
      linStru[i] += structJ[i][j];
  }

  for ( i=0; i<VarNr; i++ ) {
    colStru[i] = 0;
    for (j = 0; j < VarNr; j++)
      colStru[i] += structJ[j][i];
    colStru[i] *= linStru[i];
  }

  Stru = linStru;
  nlin = LUnonZero();
  Stru = colStru;
  ncol = LUnonZero();
  if( nlin <= ncol ) {
    Stru = linStru;
    LUnonZero();
  }
}

void BestSparsity()
{
int i,j,k;
int cnz, lnz;
int best, crt;
int best_i;
int tmp;
int s;

  UpdateStructJ();

  for ( i=0; i<VarNr; i++ )
    bestStru[i] = Code[i];

  for ( s=0; s<VarNr-1; s++ ) {
    best = MAX_SPECIES*MAX_SPECIES; best_i = 0;
    for ( i=s; i<VarNr; i++ ) {
      cnz = 0;lnz = 0;
      for (j = s; j < VarNr; j++) {
        cnz += (LUstructJ[i][j]?1:0);
        lnz += (LUstructJ[j][i]?1:0);
      }
      crt = (cnz-1)*(lnz-1);
      if( crt < best ) {
        best = crt;
        best_i = i;
      }
    }
    for ( i=0; i<VarNr; i++ ) {
      tmp = LUstructJ[s][i];
      LUstructJ[s][i] = LUstructJ[best_i][i];
      LUstructJ[best_i][i] = tmp;
    }
    for ( i=0; i<VarNr; i++ ) {
      tmp = LUstructJ[i][s];
      LUstructJ[i][s] = LUstructJ[i][best_i];
      LUstructJ[i][best_i] = tmp;
    }
    tmp = bestStru[s];
    bestStru[s] = bestStru[best_i];
    bestStru[best_i] = tmp;

    for (i = s+1; i < VarNr; i++) {
      if( LUstructJ[i][s] ) {
        for (k = s; k < VarNr; k++)
          LUstructJ[i][k] += LUstructJ[s][k];
      }
    }
  }

  Stru = bestStru;
}

void ReorderSpecies( int criteria )
{
CODE *var;
CODE *fix;
CODE *dummy;
EQ_VECT *tmpStoich_Left;
EQ_VECT *tmpStoich_Right;
EQ_VECT *tmpStoich;
CODE *tmpCode;
CODE *tmpReact;
int i, k;
int new;
int (*cmpVar)(const void *, const void *);
int (*cmpFix)(const void *, const void *);
int dummyNr;

  cmpVar = CodeCmp;
  cmpFix = CodeCmp;

  switch( criteria ) {
    case UNSORT:   cmpVar = useJacobian ? CodeRCmp : CodeCmp;
                   break;
    case LINSORT:  cmpVar = useJacobian ? CodeSCmp : CodeCmp;
                   Stru = linStru;
                   break;
    case COLSORT:  cmpVar = useJacobian ? CodeSCmp : CodeCmp;
                   Stru = colStru;
                   break;
    case BESTSORT: cmpVar = useJacobian ? NoSort : CodeCmp;
                   break;
  }

  VarNr       = 0;
  VarActiveNr = 0;
  FixNr       = 0;
  dummyNr     = 0;

  var = (CODE*)malloc( SpcNr * sizeof(CODE) );
  fix = (CODE*)malloc( SpcNr * sizeof(CODE) );
  dummy = (CODE*)malloc( 5 * sizeof(CODE) );
  tmpStoich_Left = (EQ_VECT*)malloc( SpcNr * sizeof(EQ_VECT) );
  tmpStoich_Right = (EQ_VECT*)malloc( SpcNr * sizeof(EQ_VECT) );
  tmpStoich = (EQ_VECT*)malloc( SpcNr * sizeof(EQ_VECT) );
  tmpCode = (CODE*)malloc( SpcNr * sizeof(CODE) );
  tmpReact = (CODE*)malloc( SpcNr * sizeof(CODE) );

  for( i = 0; i < SpcNr; i++ ) {
    switch( SpeciesTable[ Code[i] ].type ) {
      case VAR_SPC:  var[ VarNr++ ] = Code[ i ];
		     break;
      case FIX_SPC:  fix[ FixNr++ ] = Code[ i ];
		     break;
      case DUMMY_SPC:dummy[ dummyNr++ ] = Code[ i ];
		     break;
    }
  }

  if( Stru != bestStru ) {
    qsort( (void*)var, VarNr, sizeof(CODE), cmpVar );
  } else {
    for( i = 0; i < SpcNr; i++ )
      var[i] = bestStru[i];
  }
  qsort( (void*)fix, FixNr, sizeof(CODE), cmpFix );

  for( i = 0; i < SpcNr; i++ ) {
    EqCopy( Stoich_Left[i], tmpStoich_Left[i] );
    EqCopy( Stoich_Right[i], tmpStoich_Right[i] );
    EqCopy( Stoich[i], tmpStoich[i] );
    tmpCode[i] = Code[i];
    tmpReact[i] = Reactive[i];
  }

  SpcNr -= dummyNr;
  dummyNr = 0;

  k = 0;
  for( i = 0; i < VarNr; i++ ) {
    new = ReverseCode[ var[i] ];
    EqCopy( tmpStoich_Left[ new ], Stoich_Left[ k ] );
    EqCopy( tmpStoich_Right[ new ], Stoich_Right[ k ] );
    EqCopy( tmpStoich[ new ], Stoich[ k ] );
    Code[ k ] = tmpCode[ new ];
    Reactive[ k ] = tmpReact[ new ];
    if( Reactive[ k ] ) VarActiveNr++;
    k++;
  }
  for( i = 0; i < FixNr; i++ ) {
    new = ReverseCode[ fix[i] ];
    EqCopy( tmpStoich_Left[ new ], Stoich_Left[ k ] );
    EqCopy( tmpStoich_Right[ new ], Stoich_Right[ k ] );
    EqCopy( tmpStoich[ new ], Stoich[ k ] );
    Code[ k ] = tmpCode[ new ];
    Reactive[ k ] = tmpReact[ new ];
    k++;
  }
  for( i = 0; i < dummyNr; i++ ) {
    new = ReverseCode[ dummy[i] ];
    EqCopy( tmpStoich_Left[ new ], Stoich_Left[ k ] );
    EqCopy( tmpStoich_Right[ new ], Stoich_Right[ k ] );
    EqCopy( tmpStoich[ new ], Stoich[ k ] );
    Code[ k ] = tmpCode[ new ];
    Reactive[ k ] = tmpReact[ new ];
    k++;
  }


  for( i = 0; i < SpcNr+dummyNr; i++ )
    ReverseCode[ Code[i] ] = i;

  free( tmpReact );
  free( tmpCode );
  free( tmpStoich );
  free( tmpStoich_Right );
  free( tmpStoich_Left );
  free( dummy );
  free( fix );
  free( var );

  fflush(stdout);
}

/* Allocate Internal Arrays */
void  AllocInternalArrays( void )
{
int i;

if ( (Stoich_Left =(float**)calloc(MAX_SPECIES,sizeof(float*)))==NULL )
    FatalError(-30,"Cannot allocate Stoich_Left.\n");

for (i=0; i<MAX_SPECIES; i++)
    if ( (Stoich_Left[i] = (float*)calloc(MAX_EQN,sizeof(float)))==NULL ) {
        FatalError(-30,"Cannot allocate Stoich_Left[%d]",i,MAX_SPECIES);
    }

if ( (Stoich_Right = (float**)calloc(MAX_SPECIES,sizeof(float*)))==NULL )
    FatalError(-30,"Cannot allocate Stoich_Right.\n");

for (i=0; i<MAX_SPECIES; i++)
    if ( (Stoich_Right[i] = (float*)calloc(MAX_EQN,sizeof(float)))==NULL ) {
        FatalError(-30,"Cannot allocate Stoich_Right[%d].",i);
    }

if ( (Stoich = (float**)calloc(MAX_SPECIES,sizeof(float*)))==NULL )
    FatalError(-30,"Cannot allocate Stoich.\n");

for (i=0; i<MAX_SPECIES; i++)
    if ( (Stoich[i] = (float*)calloc(MAX_EQN,sizeof(float)))==NULL ) {
        FatalError(-30,"Cannot allocate Stoich[%d].",i);
    }

}


/* Allocate Structure Arrays */
void  AllocStructArrays( void )
{
int i;


if ( (structB = (int**)calloc(EqnNr,sizeof(int*)))==NULL )
    FatalError(-30, "Cannot allocate structB.");

for (i=0; i<EqnNr; i++)
    if ( (structB[i] =(int*) calloc(SpcNr,sizeof(int)))==NULL )
        FatalError(-30, "Cannot allocate structB[%d].\n",i);

if ( (structJ = (int**)calloc(SpcNr,sizeof(int*)))==NULL )
    FatalError(-30, "Cannot allocate structJ.");

for (i=0; i<SpcNr; i++)
    if ( (structJ[i] =(int*) calloc(SpcNr,sizeof(int)))==NULL )
        FatalError(-30, "Cannot allocate structJ[%d].\n",i);

if ( (LUstructJ = (int**)calloc(SpcNr,sizeof(int*)))==NULL )
    FatalError(-30, "Cannot allocate LUstructJ.");

for (i=0; i<SpcNr; i++)
    if ( (LUstructJ[i] = (int*)calloc(SpcNr,sizeof(int)))==NULL )
        FatalError(-30, "Cannot allocate LUstructJ[%d].\n",i);

}

/*******************************************************************/
int Postprocess( char * root )
{
char buf[ 200 ];
/*char cmd[500];*/
/*char cmdexe[500];*/
/*static char tmpfile[] = "kppfile.tmp";*/

  if ( useLang == MATLAB_LANG ) {
 /*  Add rate function definitions as internal functions to the Update_RCONST file*/
    sprintf( buf, "cat %s_Update_RCONST.m %s_Rates.m > tmp; mv tmp %s_Update_RCONST.m;",
        root, root, root );
    system( buf );
  }

/*    Postprocessing to replace parameter names by values in the declarations
  strcpy( cmd, "sed " );
  sprintf( cmd, "%s -e 's/(NVAR)/(%d)/g'", cmd, VarNr );
  sprintf( cmd, "%s -e 's/(NFIX)/(%d)/g'", cmd, FixNr );
  sprintf( cmd, "%s -e 's/(NSPEC)/(%d)/g'", cmd,SpcNr );
  sprintf( cmd, "%s -e 's/(NREACT)/(%d)/g'", cmd, EqnNr );
  sprintf( cmd, "%s -e 's/(NONZERO)/(%d)/g'", cmd, Jac_NZ );
  sprintf( cmd, "%s -e 's/(LU_NONZERO)/(%d)/g'", cmd, LU_Jac_NZ );
  sprintf( cmd, "%s -e 's/(NHESS)/(%)/g'", cmd, Hess_NZ );

  sprintf( buf, "%s_Function", rootFileName );
  switch( useLang ) {
    case F77_LANG: sprintf( buf, "%s.f", buf );
                 break;
    case F90_LANG: sprintf( buf, "%s.f90", buf );
                 break;
    case C_LANG: sprintf( buf, "%s.c", buf );
                 break;
    case MATLAB_LANG: sprintf( buf, "%s.m", buf );
                 break;
    default: printf("\n Language '%d' not implemented!\n",useLang);
                 exit(1);
  }
  sprintf( cmdexe, "%s %s > %s; mv %s %s;", cmd, buf, tmpfile, tmpfile, buf );
  printf("\n\nCMDEXE='%s'\n",cmdexe);
  system( cmdexe );
*/
}

/*******************************************************************/
int main( int argc, char * argv[] )
{
int status;
char name[ 200 ];
char *p;

  AllocInternalArrays();

  p = getenv("KPP_HOME");
  if( p ) strcpy( Home, p );

  switch( argc ) {
    case 3: eqFileName = argv[1];
            rootFileName = argv[2];
            break;
    case 2: eqFileName = argv[1];
            strcpy( name, eqFileName );
            p = name + strlen(name);
            while( p > name ) {
              if( *p == '.') {
                *p = '\0';
                break;
              }
              p--;
            }
	    rootFileName = name;
	    break;
    default: FatalError(1,"\nUsage :"
		          "\n        kpp <equations file> [output file]\n");
  }

  printf("\nThis is KPP-%s.\n", KPP_VERSION);

  printf("\nKPP is parsing the equation file.");
  status = ParseEquationFile( argv[1] );

  if( status ) FatalError(2,"%d errors and %d warnings encountered.",
                           nError, nWarning );
  /* Allocate some internal data structures */
  AllocStructArrays();

  printf("\nKPP is computing Jacobian sparsity structure.");
  ReorderSpecies( UNSORT );
  if (useReorder==1){
    BestSparsity();
    ReorderSpecies( BESTSORT );
    }
  UpdateStructJ();
  ComputeLUStructJ();

  if( initNr == -1 ) initNr = VarNr;


  printf("\nKPP is starting the code generation.");
  Generate( rootFileName );

  printf("\nKPP is starting the code post-processing.");
  Postprocess( rootFileName );

  printf("\n\nKPP has succesfully created the model \"%s\".\n\n",rootFileName);

  if( nError ) exit(4);
  if( nWarning ) exit(5);

  exit(0);
}
